R/mclosest.R

```diff
@@ -24,37 +24,39 @@
 x <- data.frame(a = 1:5, b = 3:7)
 table <- data.frame(c = c(11, 23, 3, 5, 1), d = c(32:35, 45))
 ppm <- 0.5
-#' 
+#'
 
-mclosest <- function(x, 
-                     table, 
-                     ppm = 0, 
+mclosest <- function(x,
+                     table,
+                     ppm = 0,
                      tolerance = Inf) {
   # sanity checks
-  if (is.null(dim(x)))  
+  if (is.null(dim(x)))
     stop("'x' needs to be an array")
   if (is.null(dim(table)))
     stop("'table' needs to be an array")
   if (ncol(x) != ncol(table))
     stop(" 'x' and 'table' need ot have same number of columns")
-  if (!is.matrix(x)) 
+  if (!is.matrix(x))
     x <- as.matrix(x)
-  if (!is.matrix(table)) 
+  if (!is.matrix(table))
     table <- as.matrix(table)
-  if (length(ppm) != ncol(x))
-   ppm <- rep(ppm[1], ncol(x))
-  
+  nr <- nrow(x)
+  nc <- ncol(x)
+  if (length(ppm) != nc)
+   ppm <- rep(ppm[1], nc)
+  if (length(tolerance) != nc)
+   tolerance <- rep(tolerance[1], nc)
+
   # Initialize a vector to store closest row indices
-  closest_indices <- numeric(nrow(x))
-  
-  for (i in 1:nrow(x)) { 
-    x1 <- x[i,, drop=FALSE]
-    x1 <- matrix(rbind(rep(x1, nrow(table))), ncol = ncol(table), byrow = TRUE)
-    abdiff <- abs(table - x1)
+  closest_indices <- numeric(nr)
+  for (i in seq_len(nr)) {
+    abdiff <- abs(table - rep(x[i, ], each = nr))
     ranked <- apply(abdiff, 2, rank)
+    ## Remove differences larger than tolerance
+    ranked[abdiff > rep(tolerance, each = nr)] <- NA
     rowProd <- apply(ranked, 1, prod)
     closest_indices[i] <- which.min(rowProd)
   }
-  return(closest_indices)
```

Comment:
no need to use `return` at the end of a function - only if you want to break/exit a function and still return a value `return` should be used.